### PR TITLE
a8n: Implement ExternalChangeset.diff for GitHub & Bitbucket Server

### LINKED ISSUE #6164: a8n: Display diff for changesets that exist on the codehost
This is part of https://github.com/sourcegraph/sourcegraph/issues/6085

The proposed GraphQL schema for displaying diffs for changesets that already exist on codehosts will most likely involve returning a `RepositoryComparison` for each changeset. See https://github.com/sourcegraph/sourcegraph/pull/5850 and https://github.com/sourcegraph/sourcegraph/pull/6135#issue-330763174

---

#### Backend Implementation Details

In order to implement that `diff: RepositoryComparison!` on the `a8n.changesetResolver` all that's required is — theoretically — this:

```go
func (r *changesetResolver) Diff(ctx context.Context) (*graphqlbackend.RepositoryComparisonResolver, error) {
	repo, err := r.repoResolver(ctx)
	if err != nil {
		return nil, err
	}

	base := "master"
	head := "master"

	return graphqlbackend.NewRepositoryComparison(ctx, repo, &graphqlbackend.RepositoryComparisonInput{
		Base: &base,
		Head: &head,
	})
}
```

But: how do we get to `base` and `head`? How do we get the git refs?

First answer: we can ask the codehosts.

#### GitHub

GitHub allows us to query `baseRef` and `headRef` for each pull request. But `headRef` might be null. That's the case when the pull request has been merged.

There's the `headRefOid` field, which allows us to get to the Git Object ID even if the PR has been merged and/or the commit has been deleted.

#### Bitbucket SErver

Bitbucket Server pull requests have [`FromRef` and `ToRef`](https://github.com/sourcegraph/sourcegraph/blob/f8bf51eb6543d4e5562bceaaa95465cbc52eb5c1/internal/extsvc/bitbucketserver/client.go#L755-L756) fields. The [API doesn't mention whether the `FromRef` is nullable or not](https://docs.atlassian.com/bitbucket-server/rest/6.7.1/bitbucket-rest.html#idp276).

#### Questions

Do we still want to display a diff when a changeset has been merged?

Because the next question right after "where do we get the `headRef` if it's been merged?" is: if the PR has been merged, are we sure that still have the commit in our `gitserver` copy of the repository?

I personally think that the value doesn't lie in displaying merged diff, in which case we still link to the PR on the codehost where the user can presumably still access the (cached) diff.

### PR DESCRIPTION
**Pull request by mrnugget**
This adds the ability to display diffs for changesets that already exist on Github and Bitbucket server. It fixes #6164.

The caveat, as discussed in #6164, is that we don't display a diff for changesets that have been merged or closed, since that would require changes to gitserver which are not necessary from a product's perspective (since we don't want to display diffs when a changeset has
been closed/merged)

In order to make that make that possible this makes the `diff` field on `ExternalChangeset` nullable.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 1]
```diff
@@ -692,7 +692,8 @@ type ExternalChangeset implements Changeset & Node {
     reviewState: ChangesetReviewState!
 
     # The diff of this changeset.
-    diff: RepositoryDiff!
+    # Only returned if the changeset has not been merged or closed.
+    diff: RepositoryDiff
```
[the remark this answers is not in the document]
Comment (mrnugget) [reply cmd/frontend/graphqlbackend/schema.graphql]: Hmmm, does the comment above not explain it?

> # Only returned if the changeset has not been merged or closed.

[the remark this answers is not in the document]
Comment (felixfbecker) [reply cmd/frontend/graphqlbackend/schema.graphql]: I somehow missed that 🤦🏻‍♂️

[the remark this answers is not in the document]
Comment (mrnugget) [reply cmd/frontend/graphqlbackend/schema.graphql]: Funny diff. I also had to do a double take to spot it, after thinking "wait, where's my comment?"

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 2]
```diff
@@ -638,7 +638,7 @@ interface Changeset {
     body: String!
 
     # The diff of the changeset.
```
Comment (felixfbecker) [inline cmd/frontend/graphqlbackend/schema.graphql]:
```suggestion
    # Only returned if the changeset has not been merged or closed.
```

**Pull request merged**